backend/src/llm/openai.py

```diff
@@ -98,6 +100,16 @@ async def __upload_files(self, files: list[LLMFile]) -> list[str]:
         file_ids = []
         for file in files:
             logger.info(f"Uploading file '{file.file_name}' to OpenAI")
-            file = await client.files.create(file=file.file, purpose="assistants")
+            if isinstance(file.file, (PathLike, str)):
+                file_path = Path(file.file)
+                with file_path.open("rb") as f:
+                    file_bytes = f.read()
+            elif isinstance(file.file, bytes):
+                file_bytes = file.file
+            else:
+                logger.error(f"Unsupported file type for '{file.file_name}'")
+                continue
+            file = await client.files.create(file=(file.file_name, file_bytes), purpose="assistants")
```

Comment:
Hopefully, it shouldn't be necessary to convert to `PathLike` to bytes before calling `client.files.create`. Looking at https://github.com/openai/openai-python/blob/main/src/openai/_types.py#L53 I can see that you needed to pass in the `file.file_name` as a tuple, I am hoping you can still just do this with the `PathLike` files as well? e.g.

```suggestion
             file = await client.files.create(file=(file.file_name, file.file), purpose="assistants")
```

🤞 that works as I imagine through that `f.read()` in there will slow things down

Reply:
this didn't work for me I get this error:
```
backend-1   | INFO:     172.18.0.1:47148 - "GET /suggestions HTTP/1.1" 200 OK                                                                                                                          
backend-1   | INFO:     172.18.0.4:48266 - "GET /health HTTP/1.1" 200 OK                                                                                                                               
backend-1   | INFO:     172.18.0.4:35206 - "GET /health HTTP/1.1" 200 OK
backend-1   | INFO:     Attempting to get session for session_id: a5e943a6-d811-44c3-8dfa-002840207972
backend-1   | INFO:     ***************** Session data retrieved from Redis for a5e943a6-d811-44c3-8dfa-002840207972: {}
backend-1   | INFO:     upload file type=application/pdf name=Sustainability-Report-2023.pdf size=4467363                                                                                              
backend-1   | INFO:     PDF content extracted successfully in 1.75 seconds
backend-1   | INFO:     HTTP Request: POST https://api.mistral.ai/v1/chat/completions "HTTP/1.1 200 OK"
backend-1   | INFO:     HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
backend-1   | INFO:     OpenAI response: Finish reason: stop, Content: { "files": [ "Additional-Sector-Guidance-Biotech-and-Pharma.pdf" ]}
backend-1   | INFO:     Uploading file 'Additional-Sector-Guidance-Biotech-and-Pharma.pdf' to OpenAI                                                                                                   
backend-1   | INFO:     Retrying request to /files in 0.475080 seconds
backend-1   | INFO:     Retrying request to /files in 0.859916 seconds
backend-1   | ERROR:     Connection error.
backend-1   | Traceback (most recent call last):
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1571, in _request                                                                                          
backend-1   |     response = await self._client.send(                                                                                                                                                  
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_client.py", line 1674, in send                                                                                                    
backend-1   |     response = await self._send_handling_auth(                                                                                                                                           
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_client.py", line 1702, in _send_handling_auth                                                                                     
backend-1   |     response = await self._send_handling_redirects(                                                                                                                                      
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_client.py", line 1739, in _send_handling_redirects                                                                                
backend-1   |     response = await self._send_single_request(request)                                                                                                                                  
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_client.py", line 1776, in _send_single_request                                                                                    
backend-1   |     response = await transport.handle_async_request(request)                                                                                                                             
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_transports/default.py", line 377, in handle_async_request                                                                         
backend-1   |     resp = await self._pool.handle_async_request(req)                                                                                                                                    
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/connection_pool.py", line 256, in handle_async_request                                                                   
backend-1   |     raise exc from None
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/connection_pool.py", line 236, in handle_async_request                                                                   
backend-1   |     response = await connection.handle_async_request(                                                                                                                                    
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/connection.py", line 103, in handle_async_request                                                                        
backend-1   |     return await self._connection.handle_async_request(request)                                                                                                                          
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/http11.py", line 136, in handle_async_request                                                                            
backend-1   |     raise exc
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/http11.py", line 88, in handle_async_request                                                                             
backend-1   |     await self._send_request_body(**kwargs)                                                                                                                                              
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/http11.py", line 157, in _send_request_body
backend-1   |     async for chunk in request.stream:                                                                                                                                                   
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_multipart.py", line 268, in __aiter__                                                                                             
backend-1   |     for chunk in self.iter_chunks():
backend-1   |                  ^^^^^^^^^^^^^^^^^^                                                                                                                                                      
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_multipart.py", line 230, in iter_chunks
backend-1   |     yield from field.render()                                                                                                                                                            
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_multipart.py", line 190, in render
backend-1   |     yield from self.render_data()                                                                                                                                                        
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_multipart.py", line 183, in render_data                                                                                           
backend-1   |     chunk = self.file.read(self.CHUNK_SIZE)                                                                                                                                              
backend-1   |             ^^^^^^^^^^^^^^                                                                                                                                                               
backend-1   | AttributeError: 'PosixPath' object has no attribute 'read'
backend-1   | 
backend-1   | The above exception was the direct cause of the following exception:                                                                                                                     
backend-1   |                                                                                                                                                                                          
backend-1   | Traceback (most recent call last):
backend-1   |   File "/backend/src/api/app.py", line 132, in report                                                                                                                                    
backend-1   |     processed_upload = await create_report_from_file(file)                                                                                                                               
backend-1   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/backend/src/directors/report_director.py", line 29, in create_report_from_file                                                                                                  
backend-1   |     topics = await get_materiality_agent().list_material_topics(company_name)
backend-1   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                            
backend-1   |   File "/backend/src/agents/materiality_agent.py", line 27, in list_material_topics
backend-1   |     materiality_topics = await self.llm.chat_with_file(                                                                                                                                  
backend-1   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                  
backend-1   |   File "/backend/src/llm/openai.py", line 54, in chat_with_file
backend-1   |     file_ids = await self.__upload_files(files)                                                                                                                                          
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                          
backend-1   |   File "/backend/src/llm/openai.py", line 91, in __upload_files                                                                                                                          
backend-1   |     file = await client.files.create(file=(file.file_name, file.file), purpose="assistants")
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                             
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/resources/files.py", line 422, in create
backend-1   |     return await self._post(                                                                                                                                                             
backend-1   |            ^^^^^^^^^^^^^^^^^                                                                                                                                                             
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1838, in post                                                                                              
backend-1   |     return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                       
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1532, in request                                                                                           
backend-1   |     return await self._request(                                                                                                                                                          
backend-1   |            ^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1595, in _request                                                                                          
backend-1   |     return await self._retry_request(                                                                                                                                                    
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                    
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1665, in _retry_request
backend-1   |     return await self._request(                                                                                                                                                          
backend-1   |            ^^^^^^^^^^^^^^^^^^^^                                                                                                                                                          
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1595, in _request
backend-1   |     return await self._retry_request(                                                                                                                                                    
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1665, in _retry_request                                                                                    
backend-1   |     return await self._request(                                                                                                                                                          
backend-1   |            ^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1605, in _request                                                                                          
backend-1   |     raise APIConnectionError(request=request) from err                                                                                                                                   
backend-1   | openai.APIConnectionError: Connection error.
backend-1   | 2024-12-18 16:04:35 - src.api.app - ERROR - Connection error.                                                                                                                            
backend-1   | Traceback (most recent call last):                                                                                                                                                       
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1571, in _request
backend-1   |     response = await self._client.send(                                                                                                                                                  
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                  
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_client.py", line 1674, in send
backend-1   |     response = await self._send_handling_auth(                                                                                                                                           
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                           
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_client.py", line 1702, in _send_handling_auth
backend-1   |     response = await self._send_handling_redirects(                                                                                                                                      
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                      
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_client.py", line 1739, in _send_handling_redirects
backend-1   |     response = await self._send_single_request(request)                                                                                                                                  
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_client.py", line 1776, in _send_single_request                                                                                    
backend-1   |     response = await transport.handle_async_request(request)                                                                                                                             
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_transports/default.py", line 377, in handle_async_request                                                                         
backend-1   |     resp = await self._pool.handle_async_request(req)                                                                                                                                    
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                    
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/connection_pool.py", line 256, in handle_async_request
backend-1   |     raise exc from None                                                                                                                                                                  
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/connection_pool.py", line 236, in handle_async_request                                                                   
backend-1   |     response = await connection.handle_async_request(                                                                                                                                    
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/connection.py", line 103, in handle_async_request                                                                        
backend-1   |     return await self._connection.handle_async_request(request)                                                                                                                          
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                          
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/http11.py", line 136, in handle_async_request
backend-1   |     raise exc                                                                                                                                                                            
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/http11.py", line 88, in handle_async_request                                                                             
backend-1   |     await self._send_request_body(**kwargs)
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpcore/_async/http11.py", line 157, in _send_request_body                                                                              
backend-1   |     async for chunk in request.stream:                                                                                                                                                   
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_multipart.py", line 268, in __aiter__                                                                                             
backend-1   |     for chunk in self.iter_chunks():
backend-1   |                  ^^^^^^^^^^^^^^^^^^                                                                                                                                                      
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_multipart.py", line 230, in iter_chunks                                                                                           
backend-1   |     yield from field.render()                                                                                                                                                            
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_multipart.py", line 190, in render
backend-1   |     yield from self.render_data()                                                                                                                                                        
backend-1   |   File "/usr/local/lib/python3.12/site-packages/httpx/_multipart.py", line 183, in render_data
backend-1   |     chunk = self.file.read(self.CHUNK_SIZE)                                                                                                                                              
backend-1   |             ^^^^^^^^^^^^^^                                                                                                                                                               
backend-1   | AttributeError: 'PosixPath' object has no attribute 'read'                                                                                                                               
backend-1   |                                                                                                                                                                                          
backend-1   | The above exception was the direct cause of the following exception:
backend-1   |                                                                                                                                                                                          
backend-1   | Traceback (most recent call last):                                                                                                                                                       
backend-1   |   File "/backend/src/api/app.py", line 132, in report
backend-1   |     processed_upload = await create_report_from_file(file)                                                                                                                               
backend-1   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                               
backend-1   |   File "/backend/src/directors/report_director.py", line 29, in create_report_from_file                                                                                                  
backend-1   |     topics = await get_materiality_agent().list_material_topics(company_name)                                                                                                            
backend-1   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/backend/src/agents/materiality_agent.py", line 27, in list_material_topics                                                                                                      
backend-1   |     materiality_topics = await self.llm.chat_with_file(
backend-1   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                  
backend-1   |   File "/backend/src/llm/openai.py", line 54, in chat_with_file                                                                                                                          
backend-1   |     file_ids = await self.__upload_files(files)
backend-1   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                          
backend-1   |   File "/backend/src/llm/openai.py", line 91, in __upload_files                                                                                                                          
backend-1   |     file = await client.files.create(file=(file.file_name, file.file), purpose="assistants")                                                                                             
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/resources/files.py", line 422, in create                                                                                          
backend-1   |     return await self._post(                                                                                                                                                             
backend-1   |     return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1532, in request
backend-1   |     return await self._request(
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1532, in request
backend-1   |     return await self._request(
backend-1   |     return await self._request(
backend-1   |            ^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1595, in _request
backend-1   |     return await self._retry_request(
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1665, in _retry_request
backend-1   |     return await self._request(
backend-1   |            ^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1595, in _request
backend-1   |     return await self._retry_request(
backend-1   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1665, in _retry_request
backend-1   |     return await self._request(
backend-1   |            ^^^^^^^^^^^^^^^^^^^^
backend-1   |   File "/usr/local/lib/python3.12/site-packages/openai/_base_client.py", line 1605, in _request
backend-1   |     raise APIConnectionError(request=request) from err
backend-1   | openai.APIConnectionError: Connection error.
backend-1   | INFO:     172.18.0.1:41990 - "POST /report HTTP/1.1" 500 Internal Server Error
backend-1   | INFO:     172.18.0.4:33466 - "GET /health HTTP/1.1" 200 OK
backend-1   | INFO:     172.18.0.4:57280 - "GET /health HTTP/1.1" 200 OK
backend-1   | INFO:     172.18.0.4:43876 - "GET /health HTTP/1.1" 200 OK
backend-1   | INFO:     172.18.0.4:51610 - "GET /health HTTP/1.1" 200 OK

Reply:
Ah, that's annoying, I've tried this and it seems to work:

```suggestion
            file = (file.file_name, file.file) if isinstance(file.file, bytes) else file.file
            response = await client.files.create(file=file, purpose="assistants")
```
I've changed the variable name to `response` as `file` was being overused